src/software/ai/hl/stp/skill/pass/pass_skill_fsm.h

```diff
@@ -0,0 +1,110 @@
+     *
+     * @param event the Update event
+     *
+     * @return true if the best pass so far is viable, false otherwise
+     */
+    bool passFound(const Update& event);
+
+    /**
+     * Guard that checks if the ball has been received by another friendly
+     * robot, completing the pass
+     *
+     * @param event the SuspendedUpdate event
+     *
+     * @return true if the pass was received, false otherwise
+     */
+    bool passReceived(const SuspendedUpdate& event);
+
+    /**
+     * Guard that checks if the current pass should be aborted
+     * (because the ball went astray from the target receiver point,
+     * stopped moving, got intercepted, etc.)
+     *
+     * @param event the SuspendedUpdate event
+     *
+     * @return true if the pass should be aborted, false otherwise
+     */
+    bool shouldAbortPass(const SuspendedUpdate& event);
+
+    /**
+     * Action that updates the DribbleSkillFSM to get control of the ball,
+     * while simultaneously updating the best pass so far
+     *
+     * @param event the Update event
+     * @param processEvent processes the DribbleSkillFSM::Update event
+     */
+    void findPass(const Update& event,
+                  boost::sml::back::process<DribbleSkillFSM::Update> processEvent);
+
+    /**
+     * Action that updates the PivotKickSkillFSM to take the committed pass
+     *
+     * @param event the Update event
+     * @param processEvent processes the PivotKickSkillFSM::Update event
+     */
+    void takePass(const Update& event,
+                  boost::sml::back::process<PivotKickSkillFSM::Update> processEvent);
+
+    /**
+     * Action that resets the SkillState
+     *
+     * @param event the SuspendedUpdate event
+     */
+    void resetSkillState(const SuspendedUpdate& event);
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        DEFINE_SML_STATE(DribbleSkillFSM)
+        DEFINE_SML_STATE(PivotKickSkillFSM)
+        DEFINE_SML_STATE(Suspended)
+        DEFINE_SML_EVENT(Update)
+        DEFINE_SML_EVENT(SuspendedUpdate)
+        DEFINE_SML_GUARD(passFound)
+        DEFINE_SML_GUARD(passReceived)
+        DEFINE_SML_GUARD(shouldAbortPass)
+        DEFINE_SML_ACTION(resetSkillState)
+        DEFINE_SML_SUB_FSM_UPDATE_ACTION(findPass, DribbleSkillFSM)
+        DEFINE_SML_SUB_FSM_UPDATE_ACTION(takePass, PivotKickSkillFSM)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+            *DribbleSkillFSM_S + Update_E[passFound_G] / takePass_A = PivotKickSkillFSM_S,
+            DribbleSkillFSM_S + Update_E / findPass_A,
+
+            PivotKickSkillFSM_S + Update_E / takePass_A,
+            PivotKickSkillFSM_S = Suspended_S,
+
+            Suspended_S + SuspendedUpdate_E[passReceived_G || shouldAbortPass_G] /
+                              resetSkillState_A = X,
+            Suspended_S + SuspendedUpdate_E     = Suspended_S,
+
+            X + Update_E / SET_STOP_PRIMITIVE_ACTION = X);
```

Comment:
Pass aborting was removed because I set the pass origin to always be the ball position, so as soon as we committed to a pass, the pass would be taken (no time to abort). I had to fix the pass origin to the ball position because we kept committing to a pass and then the ball would go astray from the passerPoint, causing us to overdribble the ball.

Now that we properly track dribble distances in SensorFusion and avoid overdribbling in DribbleSkillFSM, I'll set the pass origin to the passerPoint and support pass aborting.

Reply:
done